### ISSUE TITLE
CarouselWindowWidth set to 0

### ISSUE THREAD

**Issue by morinellipy** [vitalii-andriiovskyi/ngx-owl-carousel-o#82]
`this.carouselWindowWidth = this.el.nativeElement.querySelector('.owl-carousel').clientWidth;` on carousel.component.ts causes the width to be 0, therefore **slides have width 0**.

 Seen this on safari on first load and it will replicate randomly. Serving app after build for production from dist folder with http-server, **aot** on. Replicate on real device or VM as chrome emulator is not suficient.

Reason is clientWidth will return 0 for an element with display:none. 

On prebuilt-themes, .owl-carousel has display:none set.

Fixed it by setting display:block so that clientWidth returns a width different than 0

**Comment by morinellipy** [comment]
Prior comment did not work. `this.carouselWindowWidth = this.el.nativeElement.querySelector('.owl-carousel').clientWidth;` still throws 0 on safari randomly making slides to have width 0. Css class .**owl-theme** on carousel.component.ts makes sure to set display: block but this does not fix the issue.

Currently im using ViewChild to gain access to the component and use carouselService to initialize the carousel when i make sure clientWidth > 0. Something like this:
`@ViewChild('owlCar', { read: false, static: false }) owlCar` (note owlCar is not typed as i need access to private service carouselService)

`this.owlCar.carouselService.setup(this.owlCar.carouselWindowWidth, this.owlCar.slides.toArray(), this.customOptions);`
`this.owlCar.carouselService.initialize(this.owlCar.slides.toArray());`

**Comment by jumarjuaton** [comment]
How did it went @morinellipy ? Any news? I'm trying to access the component too.

**Comment by jumarjuaton** [comment]
How did you import carouselService? @morinellipy

**Comment by jumarjuaton** [comment]
@morinellipy Do you know how to trigger a refresh on the carousel?

**Comment by babulal-gaxon** [comment]
Getting same issue Width set to 0 in listing in a tab. Width is not calculating, so carousel is not showing in tab. But when i resize window its working width will be calculated and carousel will be shown there in tab.

Any can help me to resolve this issue.

**Comment by vitalii-andriiovskyi** [comment]
@babulal-gaxon  
Make first render of the carousel after the tab becomes visible. 
```HTML
<ng-container *ngIf="isTabShown">

     <owl-carousel-o [options]="customOptions" #owlCar>
         ...
     </owl-carousel-o>

  </ng-container>
```

Because it's possible the carousel renders in the case the tab isn't shown. The carousel reads the width of its parent to make further calculation and if the tab is hidden with `display: none`, it's width is 0. So the carousel get the width 0 too.

**Comment by babulal-gaxon** [comment]
Okay, Thanks for your reply ,
There is any method so i can update carousel after tab changed

**Comment by vitalii-andriiovskyi** [comment]
@babulal-gaxon 
No, there's no any methods to update carousel. Hovewer, carousel can be updated/rerendered in the case of changing the length of array with slides (removing/adding one slide) or firing the event `resize`.

**Comment by jumarjuaton** [comment]
Hello @vitalii-andriiovskyi . Is there a chance I can re-render the carousel on a div resize? This is what I get when I resize my browser:
![2020-08-27_19-54](https://user-images.githubusercontent.com/14806389/91439357-63cc5100-e89f-11ea-8b98-81479e36f49b.png)
But when I refresh, I get the correct one. Like this one:
![2020-08-27_19-54_1](https://user-images.githubusercontent.com/14806389/91439400-75adf400-e89f-11ea-8036-0e5e04d30aa2.png)

**Comment by jumarjuaton** [comment]
@vitalii-andriiovskyi Can you show an example on how to fire an event `resize`? Is this firing an event `resize` on window like this?

```
@HostListener('window:resize', ['$event'])
  sizeChange(event) {
    console.log('resize is on:' + event)
  }
```

**Comment by jumarjuaton** [comment]
My issue is that the carousel doesn't resize when I resize the browser:
![2020-08-28_11-57](https://user-images.githubusercontent.com/14806389/91519653-c831f380-e925-11ea-8f7b-5c45ece67e77.png)
But when I refresh the browser, it resizes:
![2020-08-28_11-57_1](https://user-images.githubusercontent.com/14806389/91519674-d718a600-e925-11ea-87ad-2b261f09f6ee.png)
Any thoughts @vitalii-andriiovskyi ?

**Comment by jumarjuaton** [comment]
I already got a solution. You can check this thread to get an idea: https://stackoverflow.com/questions/54016350/refresh-html-component-on-window-size-change

**Comment by vitalii-andriiovskyi** [comment]
@jumarjuaton 

using `HostListener` makes the code to listen to the event. This is not firing the event. 
I think the culprit of your problem shown in images is that DOM-element that contains `<owl-carousel>` doesn't change its width while resizing the browser..

If you resize the browser (change it's width), the event `resize` fires and carousel rerenders after that. But firstly it reads the width of its parent DOM-element and makes further calculations based on it. So check this: does the width of element, you put the carousel in, change when you resize the browser. If not, there's something wrong with CSS applied to the parent element.

**Comment by jumarjuaton** [comment]
@vitalii-andriiovskyi Yes you're right. I think it was because the parent container is in `display: flex;` Thanks for pointing me to the right direction.

**Comment by alexbliznec** [comment]
Hi @vitalii-andriiovskyi, i have the same issue. owl-carousel class has width 0 on Safari browser, template looks like this:
```<ng-container *ngIf="offers && offers.length">
                    <owl-carousel-o [options]="customOptions">
                        <ng-container *ngFor="let offer of offers; trackBy: trackByOfferId" class="tariff-slider__item">
                            <ng-template carouselSlide>
                                <div class="internet-tile">
                                    <mts-beauty-offer-tile
                                        [offer]="offer">
                                    </mts-beauty-offer-tile>
                                </div>
                            </ng-template>
                        </ng-container>
                    </owl-carousel-o>
                </ng-container>
```
So it should be displayed only if we have some cards rendered, if i understand your comment correctly.
Could you help to find decision?

**Comment by vitalii-andriiovskyi** [comment]
Hi, @alexbliznec 

You should render the carousel in the moment, when the parent tag has needed width, not 0. In your case this tag is that that holds `<ng-container *ngIf="offers && offers.length">` The carousel always reads its width and uses for own calculations and setting own width. 

By the way, using `class="tariff-slider__item"` in `<ng-container>` is useless because of this tag won't be rendered into final html.
